crates/miden-tx/src/executor/notes_checker.rs

```diff
@@ -23,17 +26,96 @@ where
+    ///
+    /// This function attempts to find the maximum set of notes that can be successfully executed
+    /// together by the target account.
+    ///
+    /// If some notes succeed but others fail, the failed notes are removed from the candidate set
+    /// and the remaining notes (successful + unattempted) are retried in the next iteration. This
+    /// process continues until either all remaining notes succeed or no notes can be successfully
+    /// executed
+    ///
+    /// # Example Execution Flow
+    ///
+    /// Given notes A, B, C, D, E:
+    /// - Try [A, B, C, D, E] → A, B succeed, C fails → Remove C, try again.
+    /// - Try [A, B, D, E] → A, B, D succeed, E fails → Remove E, try again.
+    /// - Try [A, B, D] → All succeed → Return successful=[A, B, D], failed=[C, E].
+    ///
+    /// # Returns
+    ///
+    /// Returns [`NoteConsumptionInfo`] containing:
+    /// - `successful`: Notes that can be consumed together by the account.
+    /// - `failed`: Notes that failed during execution attempts.
     pub async fn check_notes_consumability(
         &self,
         target_account_id: AccountId,
         block_ref: BlockNumber,
         input_notes: InputNotes<InputNote>,
         tx_args: TransactionArgs,
     ) -> Result<NoteConsumptionInfo, TransactionExecutorError> {
-        self.0
-            .try_execute_notes(target_account_id, block_ref, input_notes, tx_args)
+        // Ensure well-known notes are ordered first.
+        let mut notes = input_notes.into_vec();
+        notes.sort_by_key(|note| WellKnownNote::from_note(note.note()).is_none());
+
+        // Attempt to find an executable set of notes.
+        self.find_executable_notes_by_elimination(target_account_id, block_ref, notes, tx_args)
             .await
     }
+
+    /// Finds a set of executable notes and eliminates failed notes from the list in the process.
+    ///
+    /// The result contains some combination of the input notes partitioned by whether they
+    /// succeeded or failed to execute.
+    async fn find_executable_notes_by_elimination(
+        &self,
+        target_account_id: AccountId,
+        block_ref: BlockNumber,
+        notes: Vec<InputNote>,
+        tx_args: TransactionArgs,
+    ) -> Result<NoteConsumptionInfo, TransactionExecutorError> {
+        let mut candidate_notes = notes;
+        let mut failed_notes = Vec::new();
+
+        // Attempt to execute notes in a loop. Reduce the set of notes based on failures until
+        // either a set of notes executes without failure or the set of notes cannot be
+        // further reduced.
+        loop {
+            // Execute the candidate notes.
+            let execution_result = self
+                .0
+                .try_execute_notes(
+                    target_account_id,
+                    block_ref,
+                    InputNotes::<InputNote>::new_unchecked(candidate_notes.clone()),
```

Comment:
I think I would let `find_executable_notes_by_elimination` take `InputNotes` instead of `Vec<Note>`, just to communicate that these are valid input notes and not just a vector of notes. I know this will result in just calling `InputNotes::new_unchecked` in the calling function and then unwrapping it again here, but it will more clearly communicate the assumptions and why doing this `new_unchecked` here is fine.

There may even be a way to work mostly on `InputNotes` and only convert to `Vec` and back when we have to modify the inner vec (e.g. on `drain`), but maybe that's too much.